packages/expo/cli/start/server/__tests__/AsyncNgrok-test.ts

```diff
@@ -0,0 +1,158 @@
+}));
+jest.mock('../../../api/settings');
+jest.mock('../../doctor/ngrok/NgrokResolver', () => {
+  const instance: NgrokInstance = {
+    getActiveProcess: jest.fn(),
+    connect: jest.fn(async () => 'http://localhost:3000'),
+    kill: jest.fn(),
+  };
+
+  return {
+    NgrokResolver: jest.fn(() => ({
+      resolveAsync: jest.fn(async () => instance),
+      get: jest.fn(async () => instance),
+    })),
+  };
+});
+jest.mock('../../platforms/android/adbReverse', () => ({
+  startAdbReverseAsync: jest.fn(async () => true),
+}));
+jest.mock('../../../utils/exit');
+
+const asMock = <T extends (...args: any[]) => any>(fn: T): jest.MockedFunction<T> =>
+  fn as jest.MockedFunction<T>;
+
+function createNgrokInstance() {
+  const projectRoot = '/';
+  const port = 3000;
+  const ngrok = new AsyncNgrok(projectRoot, port);
+  ngrok.getActiveUrl = jest.fn(ngrok.getActiveUrl.bind(ngrok));
+  ngrok.stopAsync = jest.fn(ngrok.stopAsync.bind(ngrok));
+  return {
+    projectRoot,
+    port,
+    ngrok,
+  };
+}
+
+beforeEach(() => {
+  vol.reset();
+});
+
+describe('getActiveUrl', () => {
+  it(`is loaded on start`, async () => {
+    const { ngrok } = createNgrokInstance();
+    expect(ngrok.getActiveUrl()).toBeNull();
+    await ngrok.startAsync();
+    expect(ngrok.getActiveUrl()).toEqual('http://localhost:3000');
+  });
+});
+describe('startAsync', () => {
+  it(`fails if adb reverse doesn't work`, async () => {
+    const { ngrok } = createNgrokInstance();
+    asMock(startAdbReverseAsync).mockResolvedValueOnce(false);
+
+    await expect(ngrok.startAsync()).rejects.toThrow(/adb/);
+  });
+  it(`starts`, async () => {
+    const { ngrok } = createNgrokInstance();
+    expect(await ngrok._connectToNgrokAsync()).toEqual('http://localhost:3000');
+  });
+  it(`times out`, async () => {
+    const { ngrok } = createNgrokInstance();
+
+    // Add a connect which takes too long
+    let timer: NodeJS.Timeout | null = null;
+    const connect = jest.fn(
+      () =>
+        new Promise((resolve) => {
+          timer = setTimeout(resolve, 1000);
+        })
+    );
+    ngrok.resolver.resolveAsync = jest.fn(async () => ({ connect } as any));
+
+    try {
+      await expect(
+        ngrok._connectToNgrokAsync({
+          // Lower the time out to speed up the test.
+          timeout: 10,
+        })
+      ).rejects.toThrow(/Ngrok tunnel took too long to connect/);
+      // Time out is on a per-run basis.
+      expect(connect).toHaveBeenCalledTimes(1);
+    } finally {
+      // clean up
+      clearTimeout(timer);
+    }
+  });
+  it(`retries three times`, async () => {
+    const { ngrok } = createNgrokInstance();
+
+    // Add a connect which always fails.
+    const connect = jest.fn(() => {
+      throw new Error('woops');
+    });
+    ngrok.resolver.resolveAsync = jest.fn(async () => ({ connect } as any));
+
+    await expect(
+      ngrok._connectToNgrokAsync({
+        // Lower the time out to speed up the test.
+        timeout: 10,
```

Comment:
This is kind of OK, but take a look at Jest's fake timers. Unit tests are best when they run very fast and reliably, and having complete control over fake timers helps achieve that when we can forward the timers by 10ms or whatever amount in a programmatic way.